README.md
picoFriends is a small LibDataBroker based info addon. It provides a simple
friends display, with a tooltip that shows detailed information about friends
online.


changelog.txt
### 7.0.0.16-Release

* Tagging release version


### 7.0.0.15-Beta

* Bump for botched release to WoWI


### 7.0.0.14-Beta

* Show AFK/DND status
* Legion fixes

### 6.0.0.13-Release

* Tagging release version

### 6.0.0.12-Beta

* Export localization table to namespace
* Use icons for factions in the tooltip
* Never fail to load a client texture again
* Force a refresh when a bnet friend events fire
* Stop caching the friend data
* Show "??/??" for the default on load without data
* Refresh the summary number on load
* Nix the throttle, refresh immediately on a chat event

### 5.1.0.11-Beta

* Fix faction indicator

### 5.0.1.10-Release

* Tagging stable version

### 5.0.1.9-Beta

* Add support for battletags

### 4.2.0.8-Beta

* BNGetToonInfo returns changed

### 4.0.0.7-Beta

* Who needs friends that play starcraft?

### 4.0.0.6-Beta

* Add faction tags to realid friends

### 3.3.0.5-Beta

* Add RealID support

### 3.0.3.4-Release

* Retarded server keeps padding my notes with spaces
* LDB type

### 3.0.3.3-Beta

* tektip!
* Embed update

### 2.4.3.2-Beta

* Add support for the note field

### 2.4.2.1-Beta

* Implement LDB-1.1
* De-dongle
* Rename to picoFriends
* Add about panel


tektip.lua

local lib, oldminor = LibStub:NewLibrary("tektip-1.0", 2)
if not lib then return end
oldminor = oldminor or 0


local PADDING, VGAP, HGAP = 5, 2, 5
local TOOLTIP_PADDING = PADDING + HGAP

local bgFrame = {bgFile = "Interface\\Tooltips\\UI-Tooltip-Background", insets = {left = PADDING, right = PADDING, top = PADDING, bottom = PADDING},
	tile = true, tileSize = 16, edgeFile = "Interface\\Tooltips\\UI-Tooltip-Border", edgeSize = 16}

if oldminor < 1 then
	local linemeta = {__index = function(t,i)
		if i == "header" then
			local h = CreateFrame("Frame", nil, t.frame)
			h:SetHeight(1)
			if t.colnum == 1 then
				h:SetPoint("TOPLEFT", t.frame, "TOPLEFT", TOOLTIP_PADDING, -TOOLTIP_PADDING)
			else
				h:SetPoint("LEFT", t.frame.cols[t.colnum-1].header, "RIGHT", HGAP, 0)
			end
			t[i] = h
			return h
		end

		local fs = t.frame:CreateFontString(nil, nil, i == 1 and "GameTooltipHeaderText" or "GameTooltipText")
		if t.colnum == 1 then
			if i == 1 then
				fs:SetPoint("TOPLEFT", t.frame, "TOPLEFT", TOOLTIP_PADDING, -TOOLTIP_PADDING)
			else
				fs:SetPoint("TOPLEFT", t[i-1], "BOTTOMLEFT", 0, -VGAP)
			end
		else
			fs:SetPoint("TOP", t.frame.cols[t.colnum-1][i], "TOP")
			fs:SetPoint("LEFT", t.header, "LEFT")
			fs:SetPoint("RIGHT", t.header, "RIGHT")
		end
		fs:SetJustifyH(t.justify)
		t[i] = fs
		return fs
	end}

	function lib.new(cols, ...)
		assert(type(cols) == "number" and cols >= 2, "Must have at least two columns.")

		local f = CreateFrame("Frame", nil, UIParent)
		f:SetBackdrop(bgFrame)
		f:SetBackdropBorderColor(TOOLTIP_DEFAULT_COLOR.r, TOOLTIP_DEFAULT_COLOR.g, TOOLTIP_DEFAULT_COLOR.b)
		f:SetBackdropColor(TOOLTIP_DEFAULT_BACKGROUND_COLOR.r, TOOLTIP_DEFAULT_BACKGROUND_COLOR.g, TOOLTIP_DEFAULT_BACKGROUND_COLOR.b)
		f:SetFrameStrata("TOOLTIP")
		f:SetClampedToScreen(true)

		f.cols = {}
		for i=1,cols do
			local justify = select(i, ...) or i == 1 and "LEFT" or i == cols and "RIGHT" or "CENTER"
			f.cols[i] = setmetatable({frame = f, colnum = i, justify = justify}, linemeta)
		end

		f.AddLine = lib.AddLine
		f.AddMultiLine = lib.AddMultiLine
		f.Clear = lib.Clear
		f.AnchorTo = lib.AnchorTo

		f:SetScript("OnHide", lib.Clear)
		f:SetScript("OnShow", lib.OnShow)

		f:Hide()
		return f
	end


	local function GetTipAnchor(frame)
		local x,y = frame:GetCenter()
		if not x or not y then return "TOPLEFT", "BOTTOMLEFT" end
		local hhalf = (x > UIParent:GetWidth()*2/3) and "RIGHT" or (x < UIParent:GetWidth()/3) and "LEFT" or ""
		local vhalf = (y > UIParent:GetHeight()/2) and "TOP" or "BOTTOM"
		return vhalf..hhalf, frame, (vhalf == "TOP" and "BOTTOM" or "TOP")..hhalf
	end

	function lib:AnchorTo(frame)
		self:Clear()
		self:ClearAllPoints()
		self:SetPoint(GetTipAnchor(frame))
	end


	function lib:Clear()
		self.numlines = 0
		for i,col in pairs(self.cols) do
			for j,fs in ipairs(col) do
				fs:Hide()
			end
		end
	end


	function lib:AddLine(text, r, g, b)
		self.numlines = self.numlines + 1
		local fs = self.cols[1][self.numlines]
		fs:SetTextColor(r or NORMAL_FONT_COLOR.r, g or NORMAL_FONT_COLOR.g, b or NORMAL_FONT_COLOR.b)
		fs:SetText(text)
		fs:SetPoint("RIGHT", self.cols[#self.cols].header, "RIGHT")
		for i=2,#self.cols do self.cols[i][self.numlines]:SetText(nil) end
		fs:Show()
	end


	function lib:AddMultiLine(...)
		self.numlines = self.numlines + 1
		local numdots = select("#", ...)
		for i=1,#self.cols do
			local fs = self.cols[i][self.numlines]
			local text = select(i, ...)
			local r, g, b
			if numdots >= (#self.cols + i*3) then r, g, b = select(#self.cols + i*3 - 2, ...) end
			fs:SetTextColor(r or NORMAL_FONT_COLOR.r, g or NORMAL_FONT_COLOR.g, b or NORMAL_FONT_COLOR.b)
			fs:SetText(text == "" and " " or text)
			if i == 1 then fs:SetPoint("RIGHT", self.cols[1].header, "RIGHT") end
			fs:Show()
		end
	end
end

if oldminor < 2 then
	function lib:OnShow()
		local h, w, totalw = 0, 0, 0
		for i=1,#self.cols do
			local colw = 1
			for j=1,self.numlines do
				local fs = self.cols[i][j]
				if i == 1 then h = h + fs:GetHeight() end
				if i ~= 1 or self.cols[2][j]:GetText() then colw = math.max(colw, fs:GetStringWidth())
				else totalw = math.max(totalw, fs:GetStringWidth()) end
			end
			self.cols[i].header:SetWidth(colw)
			w = w + colw
		end
		self:SetHeight(h + TOOLTIP_PADDING*2 + (self.numlines-1)*VGAP)
		if w < (totalw - (#self.cols-1)*HGAP) then
			self:SetWidth(totalw + TOOLTIP_PADDING*2)
			local extra = totalw - w - (#self.cols-1)*HGAP
			for i=1,#self.cols do self.cols[i].header:SetWidth(extra/#self.cols + self.cols[i].header:GetWidth()) end
		else
			self:SetWidth(w + (#self.cols-1)*HGAP + TOOLTIP_PADDING*2)
		end
	end
end


tooltip.lua

local myname, ns = ...

local L = ns.L

local colors = {}
for class,c in pairs(RAID_CLASS_COLORS) do
	colors[class] = c.colorStr
end


local tip = ns.NewTooltip(4, "LEFT", "LEFT", "LEFT", "RIGHT")

local function AddDetailedLine(mylevel, level, class, name, status, note, area)
	class = class or ""
	status = status or ""
	area = area or ""
	local levelr, levelg, levelb = .5, 1, .5
	if not level then
		levelr, levelg, levelb = 1, 1, 1
		level = "|cff808080".. status.. "|r"
		status = ""
	elseif level < (mylevel - 5) then levelr, levelg, levelb = .6, .6, .6
	elseif level > (mylevel + 5) then levelr, levelg, levelb = 1, .5, .5 end
	tip:AddMultiLine(
		level or "",
		string.format(
			"|c%s%s|r%s%s",
			colors[class:gsub(" ", ""):upper()] or "ffffffff",
			name,
			status == "" and "" or " ",
			status
		),
		string.trim(note or ""),
		area,
		levelr,levelg,levelb,
		nil,nil,nil,
		1,.5,1,
		1,1,1)
end

local factiontags = {
	Horde    = " |TInterface\\TargetingFrame\\UI-PVP-Horde:0:0:0:0:64:64:0:38:0:38|t",
	Alliance = " |TInterface\\TargetingFrame\\UI-PVP-Alliance:0:0:0:0:64:64:0:38:0:38|t",
}
function ns.dataobj.OnLeave() tip:Hide() end
function ns.dataobj.OnEnter(self)
	local mylevel = UnitLevel("player")

	tip:AnchorTo(self)

	tip:AddLine("picoFriends")
	tip:AddLine(" ")

	local bnet_total, bnet_online = BNGetNumFriends()
	local wow_total, wow_online = GetNumFriends()

	for i=1,bnet_online do
		local presenceID, presenceName, battleTag, isBattleTagPresence,
		      characterName, toonID, client, online, lastOnline, isAFK, isDND,
					broadcastText, note, isFriend, broadcastTime = BNGetFriendInfo(i)
		note = note ~= "" and note
		local status = isAFK and "AFK" or isDND and "DND" or ""

		if online and toonID then
			local hasFocus, characterName, client, realmName, realmID, faction, race,
			      class, guild, area, level, gameText = BNGetGameAccountInfo(toonID)
			local texture = BNet_GetClientTexture(client)
			gameText = gameText..
			           (factiontags[faction] or "")..
			           (" |T"..texture..":0:0:0:0:64:64:4:60:4:60|t")

			AddDetailedLine(mylevel, tonumber(level), class or "", characterName,
				              status or "", note or presenceName, gameText)
		elseif online then
			tip:AddMultiLine(presenceName, client, "", nil,nil,nil , nil,nil,nil,
				               1,0,1, 1,1,1)
		end
	end

	for i = 1,GetNumFriends() do
		local name, level, class, area, connected, status, note = GetFriendInfo(i)
		if connected then
			AddDetailedLine(mylevel, level, class, name, status, note, area)
		end
	end

	if (bnet_total + wow_total) == 0 then tip:AddLine(ns.L["You have no friends!"])
	elseif (bnet_online + wow_online) == 0 then
		tip:AddLine(ns.L["No Friends Online"])
	end

	tip:Show()
end


------------------------------------------
--      Click to open friend panel      --
------------------------------------------

function ns.dataobj.OnClick()
	if FriendsFrame:IsVisible() then HideUIPanel(FriendsFrame)
	else
		ToggleFriendsFrame(1)
		FriendsFrame_Update()
		tip:Hide()
	end
end


FriendsBlock.lua

----------------------------
--      Localization      --
----------------------------

local L = {
	offline = "(.+) has gone offline.";
	online = "|Hplayer:%s|h[%s]|h has come online.";	["has come online"] = "has come online",
	["has gone offline"] = "has gone offline",

	["Level"] = "Level",
	["Name"] = "Name",
	["Emo"] = "Emo",
	["No Friends Online"] = "No Friends Online",
	["You have no friends!"] = "You have no friends!",
}


------------------------------
--      Are you local?      --
------------------------------

local friends, colors, total = {}, {}, 0
for class,color in pairs(RAID_CLASS_COLORS) do colors[class] = string.format("%02x%02x%02x", color.r*255, color.g*255, color.b*255) end


-------------------------------------------
--      Namespace and all that shit      --
-------------------------------------------

local dataobj = LibStub:GetLibrary("LibDataBroker-1.1"):NewDataObject("FriendsBlock", {icon = "Interface\\Addons\\FriendsBlock\\icon", text = "50/50"})
local f = CreateFrame("frame")
f:SetScript("OnEvent", function(self, event, ...) if self[event] then return self[event](self, event, ...) end end)


----------------------------------
--      Server query timer      --
----------------------------------

local MINDELAY, DELAY = 15, 300
local elapsed, dirty = 0, false
f:Hide()
f:SetScript("OnUpdate", function (self, el)
	elapsed = elapsed + el
	if (dirty and elapsed >= MINDELAY) or elapsed >= DELAY then ShowFriends() end
end)


local orig = ShowFriends
ShowFriends = function(...)
	elapsed, dirty = 0, false
	return orig(...)
end


----------------------
--      Enable      --
----------------------

function f:PLAYER_LOGIN()
	self:RegisterEvent("FRIENDLIST_UPDATE")
	self:RegisterEvent("CHAT_MSG_SYSTEM")

	self:Show()
	ShowFriends()

	self:UnregisterEvent("PLAYER_LOGIN")
	self.PLAYER_LOGIN = nil
end


------------------------------
--      Event Handlers      --
------------------------------

function f:CHAT_MSG_SYSTEM(event, msg)
	if string.find(msg, L["has come online"]) or string.find(msg, L["has gone offline"]) then dirty = true end
end


function f:FRIENDLIST_UPDATE()
	local online = 0
	total = 0

	local uid = GetTime()
	for i = 1,GetNumFriends() do
		local name, level, class, area, connected, status = GetFriendInfo(i)

		if name then
			if not friends[name] then friends[name] = {} end
			total = total + 1

			local t = friends[name]
			t.uid, t.level, t.class, t.area, t.status, t.connected = uid, level, class, area, status, connected
			if connected then online = online + 1 end
		end
	end

	-- Purge out deleted friends
	for name,data in pairs(friends) do if data.uid ~= uid then friends[name] = nil end end

	dataobj.text = total > 0 and string.format("%d/%d", online, total) or L["Emo"]
end


------------------------
--      Tooltip!      --
------------------------

local function GetTipAnchor(frame)
	local x,y = frame:GetCenter()
	if not x or not y then return "TOPLEFT", "BOTTOMLEFT" end
	local hhalf = (x > UIParent:GetWidth()*2/3) and "RIGHT" or (x < UIParent:GetWidth()/3) and "LEFT" or ""
	local vhalf = (y > UIParent:GetHeight()/2) and "TOP" or "BOTTOM"
	return vhalf..hhalf, frame, (vhalf == "TOP" and "BOTTOM" or "TOP")..hhalf
end


function dataobj.OnLeave() GameTooltip:Hide() end
function dataobj.OnEnter(self)
 	GameTooltip:SetOwner(self, "ANCHOR_NONE")
	GameTooltip:SetPoint(GetTipAnchor(self))
	GameTooltip:ClearLines()

	GameTooltip:AddLine("FriendsBlock")

	local online
	for name,data in pairs(friends) do
		if data.connected then
			online = true
			GameTooltip:AddDoubleLine(string.format("|cff%s%s:%s|r %s", colors[data.class:upper()] or "000000", data.level or "", name, data.status), "|cffffffff"..(data.area or ""))
		end
	end

	if total == 0 then GameTooltip:AddLine(L["You have no friends!"])
	elseif not online then GameTooltip:AddLine(L["No Friends Online"]) end

	GameTooltip:Show()
end


------------------------------------------
--      Click to open friend panel      --
------------------------------------------

function dataobj.OnClick()
	if FriendsFrame:IsVisible() then HideUIPanel(FriendsFrame)
	else
		ToggleFriendsFrame(1)
		FriendsFrame_Update()
		GameTooltip:Hide()
	end
end


-----------------------------------
--      Make rocket go now!      --
-----------------------------------

if IsLoggedIn() then f:PLAYER_LOGIN() else f:RegisterEvent("PLAYER_LOGIN") end


picoFriends.lua

local myname, ns = ...


----------------------------
--      Localization      --
----------------------------

ns.L = {
	offline = "has gone offline",
	online = "has come online",

	["Level"] = "Level",
	["Name"] = "Name",
	["Emo"] = "Emo",
	["No Friends Online"] = "No Friends Online",
	["You have no friends!"] = "You have no friends!",
}


-------------------------------------------
--      Namespace and all that shit      --
-------------------------------------------

ns.dataobj = LibStub:GetLibrary("LibDataBroker-1.1"):NewDataObject(
	"picoFriends", {
		type = "data source",
		icon = "Interface\\Addons\\picoFriends\\icon",
		text = "??/??",
})


----------------------
--      Enable      --
----------------------

function ns.OnLogin()
	ns.RegisterEvent("FRIENDLIST_UPDATE")
	ns.RegisterEvent("CHAT_MSG_SYSTEM")
	ns.RegisterEvent('BN_FRIEND_ACCOUNT_ONLINE', ns.FRIENDLIST_UPDATE)
	ns.RegisterEvent('BN_FRIEND_ACCOUNT_OFFLINE', ns.FRIENDLIST_UPDATE)
	ns.RegisterEvent('BN_FRIEND_LIST_SIZE_CHANGED', ns.FRIENDLIST_UPDATE)
	ns.RegisterEvent('BN_FRIEND_INFO_CHANGED', ns.FRIENDLIST_UPDATE)

	-- Set up the periodic refresh every 5 minutes
	-- No, I'm not passing ShowFriends directly, in case of hookers
	ns.StartRepeatingTimer(300, function() ShowFriends() end)

	-- Initialize our display and refresh data
	ns.FRIENDLIST_UPDATE()
	ShowFriends()
end


------------------------------
--      Event Handlers      --
------------------------------

function ns.CHAT_MSG_SYSTEM(event, msg)
	if string.find(msg, ns.L.online) or string.find(msg, ns.L.offline) then
		ShowFriends()
	end
end


function ns.FRIENDLIST_UPDATE()
	local bnet_total, bnet_online = BNGetNumFriends()
	local wow_total, wow_online = GetNumFriends()
	ns.dataobj.text = (bnet_total + wow_total) > 0
		and string.format("%d/%d", bnet_online + wow_online, bnet_total + wow_total)
		or ns.L["Emo"]
end
